[DNM] rbd: Replicated Write Log - team review PR

sdpeters opened the pull request:
This is all blank, because this PR will never be merged. This PR is just to talk about the code going in to https://github.com/ceph/ceph/pull/24066

sdpeters committed:
Adds Replicated Write Log, a persistent, fault-tolerant write back cache
for Ceph RBD.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Resorting to pruning out tests that fail with RWL.

Don't add RWL on images with the MIGRATING feature bit

They src and dest may have the same image name.

Also don't create image writeback at all if there is no image cache
configured.

Make deep copy test use different image names for src and dest

Fix tests for added async completions, used whether RWL is enabled or not.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
This block guard implementation isn't used anymore

Handle zero length requests

Tighten up block guard release. Remove assert for what wasn't a bug.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Skip these tests when RWL is enabled for now.

It's unclear whether they should work with the cache enabled or not.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Do a flush() before shut_down() to complete all in-flight writes and append a final sync point.

Flush to RADOS on ::flush()

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Make DiscardCopyup test not simultaneously open the same image > once

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Replaced ineffective pmem_bl copies. TestInternal passes now with RWL enabled.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
TestFUA is incompatible with RWL (multiple open of same image)

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Internal flush may be called before init succeeds

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Make RWL perf logging configurable

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Require !RWL for tests that open images > once

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Log image name and ID in key ops.

Use async completion to avoid a deadlock seen during testing

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Flush/invalidate image cache during Resize

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Disable more tests that open images > once. DeepCopy tests shoudl work on RWL, but don't.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Make invalidate after flush configurable

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
When RWL is enabled, set environment variable to skip actual PMDK persist work

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
All migration tests skipped when RWL is enabled.

Migration probably needs to disable ImageCache first.

One more DeepCopy test that dosn't work with RWL.

Tweak tests to not open images > once

Tests can close all images before fixture teardown. Avoids reopens that fil with write-back cache.

Allow saved image_ctxs to be closed by name

Close the image before re-opening it

Skip tests that reopen images when RWL is enabled

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Disable slow PMDK persistence in tests

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Consolidate constructors

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Force init, shutdown, and flush() completions to op_work_queue

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Didn't need a special case for image cache after all

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fix apparent copy-and-paste error from KernelDevice.cc

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Revert gratuitous whitespace changes

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Adds IMAGE_CACHE feature bit. Moves image cache init logic to ImageState

This is in prep. for initializing ImageCache during exclusive lock acquire.

Add types for image cache config

Can fetch image cache state during open

Record ImageCache state on image close

Init & shut down image cache in exclusive lock acquire/release

Disable checks for dirty.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Lock tests expect specific feature tests

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Remove deprecated io prio calls

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Include image-cache feature options

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Make mock tests work with IMAGE_CACHE feature on

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Make more tests work with IMAGE_CACHE feature enabled

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
More adaptation to the feature bit universe

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Reworked ImageRequestWQ<>::flush() to use ::aio_flush() again, but with a FlushSource arg.

Avoids use of flush_image(), which I think was getting blocked with the user requests in the
DeepCopy test, which was failing.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Flush to journal for internal or user flush.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Undo previous.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Add IMAGE_CACHE cases to RBD unit test

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fix apparent copy-and-paste error

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Don't build PMDK or RWL on ARM while PMDK on ARM64 is still experimental

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Skip test_librbd when features includes IMAGE_CACHE, but WITH_RWL is not defined.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Make read_only flag const, since that's how it's used

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fix message

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Don't init image cache when opening R/O unless there's a dirty one

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Add evn vars to inhibit flush and retire on close for RWL debugging

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Add ImageCache action to build

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Add env vars for testing image cache rbd tool actions

There's a regression here, but I'm committing anyway so I can migrate
to gitlab now.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fix lockdep issue in shutdown when not flushing/retiring.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Use FLUSH_SOURCE_SHUTDOWN to identify shutown flushes.

In general, a whoile lot of plumbing to get the ability to suppress RWL flushing
on close, so we can test removing and discarding the image-cache with the rbd tool.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fix rbd cli test output for new image-cache comand and --no-flush option to bench

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Clean up logging from previous test

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fix expected calls of block_writes() which now include the flush source.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Always init exclusive lock when the image cache feature is enabled.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Remove bogus assert

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Add .gitlab-ci.yml

Configure build for the available pool of runners.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

Add ctest options from environment

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

Skip run-tox* tests here

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

Fix CTEST option in .gitlab-ci.yml

Make build type RelWithDebInfo

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

Limit jobs to 4, like upstream seems to be.

Add ctest options from environment

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

Limit jobs to 4, like upstream seems to be.

sdpeters committed:
Checkpoint rbd image-cache command

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Revert "Always init exclusive lock when the image cache feature is enabled."

This reverts commit 314c82f7603f3d6eaceeee417716d09a288b4668.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fix message

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Track image-cache option changes

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fixed some things clang didn't like, which may have broken rtags

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Split WITH_PMEM into WITH_PMEM_DEVICE and WITH_PMDK

WITH_PMDK causes the PMDK (/NVML) to be built.

WITH_PMEM_DEVICE causes the PMEM OSD to be built.

WITH_PMEM_DEVICE, or WITH_RWL requires WITH_PMDK

Primordial support for WITH_PMDK_PKG enables use of PMDK
installed on platform. We'd probably like finer grained
control over this (build most of Ceph without PMDK installed,
and make PMDK an install dependency for only a few selected
things).

We can now enable RWL without requiring the PMEM OSD to build
and pass tests.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpinpdx commented on CMakeLists.txt:
```diff
@@ -267,6 +267,26 @@ if(WITH_BLUESTORE)
 
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "i386|i686|amd64|x86_64")
+  option(WITH_PMEM_DEVICE "Enable PMEM OSD" OFF)
+else()
```
Maybe WITH_PMEM_OSD would be clearer?

sdpinpdx commented on src/CMakeLists.txt:
```diff
@@ -387,6 +387,73 @@ if(WITH_DPDK)
+      DOWNLOAD_DIR ${CMAKE_BINARY_DIR}/src/
+      GIT_REPOSITORY "https://github.com/ceph/nvml.git"
+      GIT_TAG "dd622819dd4ee97d3920f913c70be"
+      SOURCE_DIR ${CMAKE_BINARY_DIR}/src/nvml
```
Just how old is this NVML tag? It might be better to express this in the form of a release tag, or at lest state what NVML release this commit should correspond to.

sdpeters committed:
Don't assert if exclusive lock release failed.

On failure, these things may not have been done.
Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Also make etags file

ctags -e, into src/TAGS.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Revert "Don't assert if exclusive lock release failed."

This reverts commit 184ba6f5e97357380d30f8012f28c34c38e6818c

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Disable almost all deep copy tests when RWL is enabled

These used to pass, and made interesting test cases, but it
probably doesn't make sense to do this with the cache enabled
anyway.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Ignore etags file TAGS

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -0,0 +1,5130 @@
+      C_WriteRequest<T>::finish_req(r);
+    } else {
+      utime_t now = ceph_clock_now();
+
```
I don't think we need to check whether m_perfcounter is null here. During init, it is initialized and isn't checked in other places.

lixiaoy1 commented on src/librbd/cache/ReplicatedWriteLog.cc:
```diff
@@ -0,0 +1,5130 @@
+      C_WriteRequest<T>::finish_req(r);
+    } else {
+      utime_t now = ceph_clock_now();
+
```
I don't think this check (m_perfcount is null not or ) is needed here. It is initialized and no check in other places.

lixiaoy1 replied:
I added the comments at line 4293 in ReplicatedWriteLog<I>::flush() , not sure why the display here is line 1788.

sdpeters replied:
> I added the comments at line 4293 in ReplicatedWriteLog::flush() , not sure why the display here is line 1788.

It’s shown with the wrong line here, but when you open the diffs for the file it’s on the right line.  It may be unhappy about files this large, or about the updates I’ve pushed since this PR was created.

sdpeters replied:
> I don't think this check (m_perfcount is null not or ) is needed here. It is initialized and no check in other places.

The idea was for the perf counters to be optional. If there’s no way to skip creating the perf counter object, there should be. It may only be a build option, but it should be a runtime option. I’m not worried about the performance effect of reading a pointer here. The object memory allocation overhead is much larger than that, and more variable. If we were worried about this, we could make the perfcounter a template argument (const) so the compiler would generate a version of the function with the perf counter clause, and one without.

lixiaoy1 replied:
I agree that if it is optional, it should be a runtime option. It means customers can change the option when the service is running. I proposed the above comments as I think we should use logger consistently in RWL. Currently, perf is always enabled, so no need check. 
To make perfcounter a template argument, can it be changed in runtime?
For reference, there is perf and perf_rocksdb options in common/options.cc, but I don't find the usage of perf. Meanwhile, perf_rocksdb is only used for rocksdb inside stats CreateDBStatistics, not used for controlling perfcounter.

sdpeters replied:
I remember now. This is because the ImageCache::flush() method might be called by librbd before the cache initializes, or when initialization fails. If you take that out, you should see a segfault near there during the librbd unit test.

sdpeters committed:
Adds ReplicatedWriteLogInternal.h

Hopefully trims enough off of ReplicatedWriteLog.cc that gitlab will
show the diffs in a merge request.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Shift more things from RWL.cc to RWLInternal.h

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Finish moving lines

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
One more try for gitlab-friendly file sizes.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

lixiaoy1 committed:
To fix IO hange problem caused by aio_flush.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

lixiaoy1 committed:
Add IO test for RWL

It does IO test with RWL, and tries to cover the test case when
RWL cache is full. But it is not guranteed.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

lixiaoy1 committed:
Fix the bug 'First IO hangs if RWL can't be initialized'.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

lixiaoy1 committed:
Update default and minimum value of rwl_size.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

Commit:
Fixes the image-cache show command

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

lixiaoy1 committed:
Update RWL full case to use default value.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

lixiaoy1 committed:
Add pool name in the name of rwl cache file.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fixed rebase issues.

Look closer at whether FLUSH_SOURCE_WRITEBACK should be a no-op in
RWL, or an assert.

Look closer at discard granularity passed to lower cache. This replaces
skip_partial. We were always passing skip_partial = false, so the RWL
and the image got the same discard results (regions which read as zero,
vs. whatever was there before).

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Fix expected usage output

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Probably rebase noise

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Repeat compare buffer as necessary to match length of read in compare-and-write

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Disable some more migration tests when IMAGE_CACHE feature is enabled.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

Commit:
fix image-cache disable feature command

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

lixiaoy1 committed:
Start rwl threads after rwl_init.

Signed-off-by: Li Xiaoyan <xiaoyan.li@intel.com>

yison committed:
change ENOENT to EPERM when open image cache failed;

Signed-off-by: Yuan Lu <yuan.y.lu@intel.com>

Commit:
image-cache discard-dirty CLI command

Signed-off-by: Mahati Chamarthy <mahati.chamarthy@intel.com>

yison committed:
fixed issue #35: If the image metadata says there is no cache, any existing pool file with the same name should be deleted.

Signed-off-by: Yuan Lu <yuan.y.lu@intel.com>

yison committed:
only image cache state.present=0 and has pool file, then del the pool file with the same name;

Signed-off-by: Yuan Lu <yuan.y.lu@intel.com>

yison committed:
do this in rwl_int() function;

Signed-off-by: Yuan Lu <yuan.y.lu@intel.com>

yison committed:
fix #41:IO hang when rwl_size = 10G or 20G and without rate_iops limit, While there's no IO problem when rwl_size=1G

Signed-off-by: Yuan Lu <yuan.y.lu@intel.com>

sdpeters committed:
Pass through everything to writeback interface

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

yison committed:
add high/low water log entries threshold and trigger it to retire entries when there's no enough log entries;

Signed-off-by: Yuan Lu <yuan.y.lu@intel.com>

yison committed:
clean indent issue;

Signed-off-by: Yuan Lu <yuan.y.lu@intel.com>

yison committed:
modify indentation based on Scott's feedback;

Signed-off-by: Yuan Lu <yuan.y.lu@intel.com>

lixiaoy1 committed:
Fix the issue of flushing rwl cache when showing image cache.

Signed-off-by: Xiaoyan Li <xiaoyan.li@intel.com>

lixiaoy1 committed:
Fix image-cache show issue.

Signed-off-by: Xiaoyan Li <xiaoyan.li@intel.com>

lixiaoy1 committed:
pmemobj_close won't fail.

Signed-off-by: Xiaoyan Li <xiaoyan.li@intel.com>

sdpeters committed:
Rebase repairs. Mostly trivial except async op handling in flush.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Make imageFlushRequest::send_request() more like it was before the rebase.

All in pursuit of issue #48 after the latest rebase.

The aio_comp() operations in comments there cause an assert in init or open now.

FLUSH_SOURCE_WRITEBACK requests complete synchonously now, like they did before the rebase.

None of this fixes the test assert in issue #48.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Replace namespace declaration that is present in master but not our branch

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Supplied initial values for things not initialized by all constructors

I saw compiler warnings about these fields being used uninitialized and thought it
might be related to the rebase regression (it probably isn't).

It does look like these could be left uninitialized if the ReplicatedWriteLogSpec
constructor with no argumets was used.

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

sdpeters committed:
Bad merge

Signed-off-by: Scott Peterson <scott.d.peterson@intel.com>

lixiaoy1 committed:
Fix unit test problem.

Signed-off-by: Xiaoyan Li <xiaoyan.li@intel.com>

sdpeters committed:
Update README.md
